src/pkg/zoci/copier.go

```diff
@@ -5,53 +5,83 @@
 package zoci
 
 import (
-	"bytes"
 	"context"
+	"fmt"
+	"time"
 
-	"github.com/defenseunicorns/pkg/oci"
-	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/zarf-dev/zarf/src/internal/packager/images"
 	"github.com/zarf-dev/zarf/src/pkg/logger"
-	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2"
+
+	retry "github.com/avast/retry-go/v4"
 )
 
-// CopyPackage copies a zarf package from one OCI registry to another
-func CopyPackage(ctx context.Context, src *Remote, dst *Remote, concurrency int) (err error) {
+// CopyPackage copies a zarf package from one OCI registry to another using ORAS with retry.
+func CopyPackage(ctx context.Context, src *Remote, dst *Remote, concurrency int, retries int) (err error) {
 	l := logger.From(ctx)
 	if concurrency <= 0 {
 		concurrency = DefaultConcurrency
 	}
 
-	srcManifest, err := src.FetchRoot(ctx)
-	if err != nil {
-		return err
-	}
-	l.Info("copying package",
-		"src", src.Repo().Reference.String(),
-		"dst", dst.Repo().Reference.String())
-	if err := oci.Copy(ctx, src.OrasRemote, dst.OrasRemote, nil, concurrency, nil); err != nil {
-		return err
-	}
-
+	// Resolve the root digest of the source package (manifest or index)
 	srcRoot, err := src.ResolveRoot(ctx)
 	if err != nil {
 		return err
 	}
+	srcRef := srcRoot.Digest.String()
 
-	b, err := srcManifest.MarshalJSON()
-	if err != nil {
-		return err
-	}
-	expected := content.NewDescriptorFromBytes(ocispec.MediaTypeImageManifest, b)
+	copyOpts := dst.OrasRemote.GetDefaultCopyOpts()
+	copyOpts.Concurrency = concurrency
 
-	if err := dst.Repo().Manifests().PushReference(ctx, expected, bytes.NewReader(b), srcRoot.Digest.String()); err != nil {
-		return err
-	}
+	tag := src.Repo().Reference.Reference // keep the source tag on the destination
 
-	tag := src.Repo().Reference.Reference
-	if err := dst.UpdateIndex(ctx, tag, expected); err != nil {
-		return err
+	err = retry.Do(
+		func() error {
+			l.Info("copying package",
+				"src", src.Repo().Reference.String(),
+				"dst", dst.Repo().Reference.String(),
+				"ref", srcRef,
+			)
+			source := src.Repo() // implements oras.ReadOnlyTarget
+			trackedDst := images.NewTrackedTarget(
+				dst.Repo(),
+				0, // unknown total for registry→registry copy
```

Comment:
Maybe adding progress here should be a separate PR, currently this would not report progress. https://github.com/zarf-dev/zarf/blob/8067c8b31deea98610917cdcd918ede5ac0d20af/src/internal/packager/images/progress.go#L27-L30

It should be possible to get all the descs from the mainfest though then sum the total size.

Reply:
I agree - did some looking at what we need and do feel it extends scope and would be well fit for a future PR given parity.